nibabel/freesurfer/io.py

```diff
@@ -318,51 +338,73 @@ def read_annot(filepath, orig_ids=False):
+
+        # in old-format files, the next field will contain the number of
+        # entries in the colour table. In new-format files, this will be
+        # equal to -2
         n_entries = np.fromfile(fobj, dt, 1)[0]
+
+        # We've got an old-format .annot file.
         if n_entries > 0:
+
+            # orig_tab string length + string
             length = np.fromfile(fobj, dt, 1)[0]
             orig_tab = np.fromfile(fobj, '>c', length)
             orig_tab = orig_tab[:-1]
-
             names = list()
-            ctab = np.zeros((n_entries, 5), np.int)
+            ctab = np.zeros((n_entries, 5), dt)
             for i in xrange(n_entries):
+                # structure name length + string
                 name_length = np.fromfile(fobj, dt, 1)[0]
                 name = np.fromfile(fobj, "|S%d" % name_length, 1)[0]
                 names.append(name)
+                # read RGBA for this entry
                 ctab[i, :4] = np.fromfile(fobj, dt, 4)
-                ctab[i, 4] = (ctab[i, 0] + ctab[i, 1] * (2 ** 8) +
-                              ctab[i, 2] * (2 ** 16) +
-                              ctab[i, 3] * (2 ** 24))
+        # We've got a new-format .annot file
         else:
+            # file version number
             ctab_version = -n_entries
             if ctab_version != 2:
                 raise Exception('Color table version not supported')
-            n_entries = np.fromfile(fobj, dt, 1)[0]
-            ctab = np.zeros((n_entries, 5), np.int)
+            # maximum LUT index present in the file
+            max_index = np.fromfile(fobj, dt, 1)[0]
+            ctab = np.zeros((max_index, 5), dt)
+            # orig_tab string length + string
             length = np.fromfile(fobj, dt, 1)[0]
             np.fromfile(fobj, "|S%d" % length, 1)[0]  # Orig table path
+            # number of LUT entries present in the file
             entries_to_read = np.fromfile(fobj, dt, 1)[0]
             names = list()
-            for i in xrange(entries_to_read):
-                np.fromfile(fobj, dt, 1)[0]  # Structure
+            for _ in xrange(entries_to_read):
+                # index of this entry
+                idx = np.fromfile(fobj, dt, 1)[0]
+                # structure name length + string
                 name_length = np.fromfile(fobj, dt, 1)[0]
                 name = np.fromfile(fobj, "|S%d" % name_length, 1)[0]
                 names.append(name)
-                ctab[i, :4] = np.fromfile(fobj, dt, 4)
-                ctab[i, 4] = (ctab[i, 0] + ctab[i, 1] * (2 ** 8) +
-                              ctab[i, 2] * (2 ** 16))
-        ctab[:, 3] = 255
+                # RGBA
+                ctab[idx, :4] = np.fromfile(fobj, dt, 4)
 
-    labels = labels.astype(np.int)
+    # generate annotation values for each LUT entry
+    ctab[:, [4]] = _pack_rgba(ctab[:, :4])
+
+    # make sure names are strings, not bytes
+    names = [n.decode('ascii') for n in names]
```

Comment:
Names are always ascii?  Or can be latin1?

Reply:
Not entirely sure. Latin1 is a superset of ascii, right? So it should be no harm to change to latin1.

 - https://github.com/freesurfer/freesurfer/blob/dev/matlab/read_annotation.m#L167
 - https://github.com/freesurfer/freesurfer/blob/dev/matlab/write_annotation.m#L118

Reply:
Oh, forgot about this issue. Makes sense to me to be very liberal about reading and accept any byte string at all.

First, it looks like we've returning `str`/`bytes` in Python 2/3, right? So decoding to `unicode`/`str` would be an API change. Given that, I think it makes sense to continue returning `bytes` by default. If you want to add in a `name_encoding` (or similar) parameter, you could do:

```Python
if name_encoding is not None:
    names = [n.decode(name_encoding) for n in names]
```

Also, looking at those MATLAB files, there's some NUL-termination we're not handling correctly on the writing end. (Numpy correctly truncates null characters.) I'll put a comment in `write_annot` where I think we need to fix that.

Reply:
Leave the return value as str/bytes for 2/3 - fair enough. I made this change to bring the code in line with the function docstring, which promises a `list of str`(I'm guessing it was written against python 2 only, so was not handling python 3 at all).

Reply:
Yes, that's almost certainly true. Probably good to update the docstring to say `bytes`, then.